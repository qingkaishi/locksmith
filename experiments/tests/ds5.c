/* Data structure example 5:  A recursive data structure.  All
   elements of the list are protected by the lock in that element.
   Will report a race since the lock will be non-linear. */
#include <pthread.h>

struct list {
  int data;
  pthread_mutex_t lock;
  struct list *next;
};

struct list *shared = NULL;

void *run(void *arg) {
  struct list *cur;

  for (cur = shared; cur; cur = cur->next) {
    pthread_mutex_lock(&cur->lock);
    cur->data++;
    pthread_mutex_unlock(&cur->lock);
  }

	cur = cur->next;
  return NULL;
}
int main() {
  pthread_t t1, t2;
  int i;

  for (i = 0; i < 42; i++) {
    struct list *new;
	  new = (struct list *) malloc(sizeof(struct list));
    new->next = shared;
    pthread_mutex_init(&new->lock, NULL);
    shared = new;
  }
  for (i = 0; i < 42; i++) {
    struct list *new = (struct list *) malloc(sizeof(struct list));
    new->next = shared;
    pthread_mutex_init(&new->lock, NULL);
    shared = new;
  }
  pthread_create(&t1, NULL, run, NULL);
  pthread_create(&t2, NULL, run, NULL);
  return 1;
}
