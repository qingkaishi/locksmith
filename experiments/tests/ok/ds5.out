************************* STARTING *************************

shared: alloc.data:ds5.c:32
        alloc.data:ds5.c:38

Warning: Possible data race: alloc.data:ds5.c:38 is protected by non-linear or concrete lock(s):
  *cur.lock:ds5.c:12
  *new.lock:ds5.c:31
  *new___0.lock:ds5.c:38
  non linear concrete new->lock:ds5.c:34
  non linear concrete new___0->lock:ds5.c:40
 references:
  dereference of *cur.data:ds5.c:12 at ds5.c:19
    alloc.data:ds5.c:38 => *cur.data:ds5.c:12
  locks acquired:
    *cur.lock:ds5.c:12
    *new.lock:ds5.c:31
    *new___0.lock:ds5.c:38
    non linear concrete new->lock:ds5.c:34
    non linear concrete new___0->lock:ds5.c:40
  in: main at ds5.c:26 



Warning: Possible data race: alloc.data:ds5.c:32 is protected by non-linear or concrete lock(s):
  *cur.lock:ds5.c:12
  *new.lock:ds5.c:31
  *new___0.lock:ds5.c:38
  non linear concrete new->lock:ds5.c:34
  non linear concrete new___0->lock:ds5.c:40
 references:
  dereference of *cur.data:ds5.c:12 at ds5.c:19
    alloc.data:ds5.c:32 => *cur.data:ds5.c:12
  locks acquired:
    *cur.lock:ds5.c:12
    *new.lock:ds5.c:31
    *new___0.lock:ds5.c:38
    non linear concrete new->lock:ds5.c:34
    non linear concrete new___0->lock:ds5.c:40
  in: main at ds5.c:26 


*************************** DONE ***************************
